subt/main.py

```diff
@@ -545,16 +547,17 @@ def on_acc(self, timestamp, data):
                 raise Collision()
 
     def on_artf(self, timestamp, data):
-        if self.offset is None:
+        if self.offset is None or self.orientation is None or self.xyz is None:
             # there can be observed artifact (false) on the start before the coordinate system is defined
             return
-        artifact_data, deg_100th, dist_mm = data
+        artifact_data, dx, dy, dz = data
+        vector = [dx, dy, dz]
+        dx, dy, dz = quaternion.rotate_vector(vector, self.orientation)
         x, y, z = self.xyz
         x0, y0, z0 = self.offset
-        angle, dist = self.yaw + math.radians(deg_100th / 100.0), dist_mm / 1000.0
-        ax = x0 + x + math.cos(angle) * dist
-        ay = y0 + y + math.sin(angle) * dist
-        az = z0 + z
+        ax = x0 + x + dx/1000.0
+        ay = y0 + y + dy/1000.0
+        az = z0 + z + dz/1000.0
```

Comment:
`self.xyz` is set in `on_pose3d` and I believe it is already corrected for origin offset (it either comes from `odoimuloc` or from `offseter`).

Reply:
Note, that I did not change this part - and offseter is not plugged in as input ... only "independent reference" (as discussed over email). I.e. so far I keep the "compatibility" here.

Reply:
self.xyz used to come only from pose2d which has always been without origin offset
pose3d was from the begining with origin offset but was not used for anything inside app
somewhere along the line on_pose3d started to set self.xyz as well at which point it become mangled and beyond my capacity to comprehend